File: hash/swift/RosettaImageHash/Sources/RosettaImageHash/ColorHash.swift
import Foundation

/// colorhash: HSV-binned histogram hash.
///
/// Bins: black (L<32), gray (L>=32 and S<85), 6 faint hue bins (85<=S<170),
/// 6 bright hue bins (S>170). S==170 increments colorfulCount (denominator)
/// but lands in neither hue histogram.
public func colorhash(_ image: RGBImage, binbits: Int) throws -> Hash {
    guard binbits >= 1 else { throw ImageHashError.invalidBinbits(binbits) }
    let rgb = ImgRGB.toRGB(image)
    let data = rgb.data
    let w = rgb.width
    let h = rgb.height
    let n = w * h

    var blackCount = 0
    var grayCount = 0
    var colorfulCount = 0
    var faintBins = [0, 0, 0, 0, 0, 0]
    var brightBins = [0, 0, 0, 0, 0, 0]

    var si = 0
    for _ in 0..<n {
        let r = Int(data[si])
        let g = Int(data[si + 1])
        let b = Int(data[si + 2])
        si += 3
        let l = toGray(r, g, b)
        if l < 32 {
            blackCount += 1
            continue
        }
        let (hue, s, _) = toHSV(r, g, b)
        if s < 85 {
            grayCount += 1
            continue
        }
        colorfulCount += 1
        var hueBin = (hue * 6) / 255
        if hueBin > 5 { hueBin = 5 }
        if s < 170 {
            faintBins[hueBin] += 1
        } else if s > 170 {
            brightBins[hueBin] += 1
        }
    }

    let maxVal = 1 << binbits
    let c = max(1, colorfulCount)
    let clip: (Int) -> Int = { v in v > maxVal - 1 ? maxVal - 1 : v }

    var values = [Int](repeating: 0, count: 14)
    values[0] = clip((blackCount * maxVal) / n)
    values[1] = clip((grayCount * maxVal) / n)
    for i in 0..<6 {
        values[2 + i] = clip((faintBins[i] * maxVal) / c)
        values[8 + i] = clip((brightBins[i] * maxVal) / c)
    }

    var bits: [[Bool]] = []
    for i in 0..<14 {
        bits.append(colorhashBinEncode(values[i], binbits: binbits))
    }
    return try Hash(bits: bits)
}

/// SPEC.md §8 quirky bin encoding.
///
/// bit[i] = (v >> (B-i-1)) & ((1 << (B-i)) - 1) > 0
///
/// Worked: v=4, B=4 → [F,T,T,F] (0x6); v=8, B=4 → [T,T,F,F] (0xc). NOT standard binary.
public func colorhashBinEncode(_ v: Int, binbits: Int) -> [Bool] {
    var bits = [Bool](repeating: false, count: binbits)
    for i in 0..<binbits {
        let shifted = v >> (binbits - i - 1)
        let masked = shifted & ((1 << (binbits - i)) - 1)
        bits[i] = masked > 0
    }
    return bits
}

File: hash/java/src/main/java/io/github/wmetcalf/rosettasquint/hash/ColorHash.java
package io.github.wmetcalf.rosettasquint.hash;

import io.github.wmetcalf.rosettasquint.hash.internal.BufferedImageRgb;
import io.github.wmetcalf.rosettasquint.hash.internal.PilGrayscale;
import io.github.wmetcalf.rosettasquint.hash.internal.PilHsv;

import java.awt.image.BufferedImage;

/**
 * colorhash: HSV-binned histogram hash. Bins:
 *   - black: L < 32
 *   - gray: L >= 32 and S < 85
 *   - 6 faint hue bins: L >= 32 and 85 <= S < 170
 *   - 6 bright hue bins: L >= 32 and S > 170
 * Quirky bin encoding (SPEC.md §8) used to pack each bin's count into binbits bits.
 */
public final class ColorHash {
    public static ImageHash compute(BufferedImage img) {
        return compute(img, 3);
    }

    public static ImageHash compute(BufferedImage img, int binbits) {
        if (binbits < 1)
            throw new IllegalArgumentException("binbits must be >= 1");
        // (1 << binbits) overflows for large binbits. Cross-port practical
        // limit is 30 (JS bitwise int range), so cap there for parity.
        // Real use is binbits 3-8.
        if (binbits > 30)
            throw new IllegalArgumentException("binbits must be <= 30");

        int[][] rgb = BufferedImageRgb.toIntArray(img);
        int h = rgb.length;
        int w = rgb[0].length;
        long n = (long) w * h;

        long blackCount = 0;
        long grayCount = 0;
        long colorfulCount = 0;
        long[] faintBins = new long[6];
        long[] brightBins = new long[6];

        for (int y = 0; y < h; y++) {
            for (int x = 0; x < w; x++) {
                int p = rgb[y][x];
                int r = (p >> 16) & 0xFF, g = (p >> 8) & 0xFF, b = p & 0xFF;
                int l = PilGrayscale.toGray(r, g, b);
                if (l < 32) {
                    blackCount++;
                    continue;
                }
                int[] hsv = PilHsv.toHsv(r, g, b);
                int s = hsv[1];
                if (s < 85) {
                    grayCount++;
                    continue;
                }
                colorfulCount++;
                int hueBin = Math.min(5, (int) (hsv[0] * 6.0 / 255.0));
                if (s < 170) {
                    faintBins[hueBin]++;
                } else if (s > 170) {
                    brightBins[hueBin]++;
                }
            }
        }

        int maxVal = 1 << binbits;
        long c = Math.max(1, colorfulCount);
        int[] values = new int[14];
        values[0] = Math.min(maxVal - 1, (int) ((double) blackCount / n * maxVal));
        values[1] = Math.min(maxVal - 1, (int) ((double) grayCount / n * maxVal));
        for (int i = 0; i < 6; i++) {
            values[2 + i] = Math.min(maxVal - 1, (int) ((double) faintBins[i] * maxVal / c));
            values[8 + i] = Math.min(maxVal - 1, (int) ((double) brightBins[i] * maxVal / c));
        }

        boolean[][] bits = new boolean[14][binbits];
        for (int i = 0; i < 14; i++) {
            boolean[] vbits = binEncode(values[i], binbits);
            System.arraycopy(vbits, 0, bits[i], 0, binbits);
        }
        return new ImageHash(bits);
    }

    /**
     * SPEC.md §8 quirky bin encoding. Verified: v=4 -> 0110, v=8 -> 1100 (not standard binary).
     */
    public static boolean[] binEncode(int v, int binbits) {
        boolean[] bits = new boolean[binbits];
        for (int i = 0; i < binbits; i++) {
            int shifted = v >>> (binbits - i - 1);
            int masked = shifted & ((1 << (binbits - i)) - 1);
            bits[i] = masked > 0;
        }
        return bits;
    }

    private ColorHash() {}
}
